File: godart.go
package main

import (
	"C"
	"fmt"
	"github.com/mraleph/go_dart_ffi_example/dart_api_dl"
	"time"
	"unsafe"
)

//export InitializeDartApi
func InitializeDartApi(api unsafe.Pointer) {
	dart_api_dl.Init(api)
}

//export StartWork
func StartWork(port int64) {
	fmt.Println("Go: Starting some asynchronous work")
	go work(port)
	fmt.Println("Go: Returning to Dart")
}

func work(port int64) {
	var counter int64
	for {
		time.Sleep(2 * time.Second)
		fmt.Println("GO: 2 seconds passed")
		counter++
		dart_api_dl.SendToPort(port, counter)
	}
}

// Unused
func main() {}

File: godart.dart
import 'dart:ffi';
import 'dart:isolate';

typedef StartWorkType = Void Function(Int64 port);
typedef StartWorkFunc = void Function(int port);

void main() async {
  final lib = DynamicLibrary.open("./godart.so");

  final initializeApi = lib.lookupFunction<IntPtr Function(Pointer<Void>),
      int Function(Pointer<Void>)>("InitializeDartApi");
  if (initializeApi(NativeApi.initializeApiDLData) != 0) {
    throw "Failed to initialize Dart API";
  }

  final interactiveCppRequests = ReceivePort()
    ..listen((data) {
      print('Received: ${data} from Go');
    });
  final int nativePort = interactiveCppRequests.sendPort.nativePort;

  final StartWorkFunc startWork =
      lib.lookup<NativeFunction<StartWorkType>>("StartWork").asFunction();
  startWork(nativePort);
  while (true) {
    await Future.delayed(const Duration(seconds: 2));
    print("Dart: 2 seconds passed");
  }
}
